### PR TITLE
Add support for windows to Karate tests

### PR DESCRIPTION
**Pull request by GabrielFleischer**
This pull request allows Windows users to run karate tests (at least for backends)

It was tested on both windows and WSL (Ubuntu) and the tests were running.

### CODE REVIEW

File: tests/karate/src/test/java/be/utils/ScalaServer.java [hunk 1]
```diff
@@ -23,9 +25,25 @@ public void stop() {
...
-  public String[] getCmd() {
+  public String[] getCmd() throws IOException {
     String configPath = Paths.get("src", "main", "scala", "ch", "epfl", "pop", "config").toString();
-    return new String[] { "sbt", "-Dscala.config=" + configPath, "run" };
+    File targetJar = getTargetJar();
+
+    return new String[] { "java", "-Dscala.config=" + configPath, "-jar", targetJar.getCanonicalPath()};
+  }
+
+  private File getTargetJar() throws IOException {
```
Comment (jbsv) [inline tests/karate/src/test/java/be/utils/ScalaServer.java:35]: Maybe more vertical spacing to improve readability ;-)

Comment (GabrielFleischer) [reply tests/karate/src/test/java/be/utils/ScalaServer.java:38]: Fixed

**Review by jbsv — APPROVED** [review]

File: tests/karate/src/test/java/be/utils/Server.java [hunk 2]
```diff
@@ -32,9 +38,33 @@ private static ProcessBuilder build(String[] cmd, String dir, String logPath) {
...
 
   @Override
-  public void stop() {
-    process.children().forEach(ProcessHandle::destroyForcibly);
+  public void stop() throws RuntimeException {
+    process.descendants().forEach(ProcessHandle::destroy);
+    process.destroy();
+    List<CompletableFuture<?>> allProcesses = process.descendants().map(ProcessHandle::onExit).collect(Collectors.toList());
+    allProcesses.add(process.onExit());
+    CompletableFuture<Void> onExit = CompletableFuture.allOf(allProcesses.toArray(new CompletableFuture[]{}));
```
Comment (pierluca) [inline tests/karate/src/test/java/be/utils/Server.java:46]: this thing looks a bit cryptic to me. Can we improve the readability ?

```suggestion
    List<CompletableFuture<?>> allProcesses = process
      .descendants()
      .map(ProcessHandle::onExit)
      .collect(Collectors.toList());

    allProcesses.add(process.onExit());

    CompletableFuture<Void> onExit = CompletableFuture
      .allOf(allProcesses.toArray(new CompletableFuture[]{}));
```

File: tests/karate/src/test/java/be/utils/ScalaServer.java (hunk 1 above)
Comment (GabrielFleischer) [reply tests/karate/src/test/java/be/utils/ScalaServer.java:38]: Did my commit fail... ?

File: tests/karate/src/test/java/be/utils/server.feature [hunk 3]
```diff
@@ -96,6 +96,7 @@ Feature: This feature starts a server and stops it after every scenario.
           """
             function() {
               stopServer();
+              wait(2);
```
[the remark this answers is not in the document]
Comment (GabrielFleischer) [reply tests/karate/src/test/java/be/utils/server.feature:99]: It wasn't the case before this PR.
Now it should be the case, this is an artifact of my debugging.

Thanks for spotting it

File: tests/karate/src/test/java/be/utils/Server.java (hunk 2 above)
Comment (GabrielFleischer) [reply tests/karate/src/test/java/be/utils/Server.java:46]: Sure

File: tests/karate/src/test/java/be/utils/Server.java [hunk 4]
```diff
@@ -28,32 +28,48 @@ private static ProcessBuilder build(String[] cmd, String dir, String logPath) {
...
+      .allOf(allProcesses.toArray(new CompletableFuture[]{}));
 
     try {
       // Wait for the process to exit
       onExit.get(1, TimeUnit.SECONDS);
     } catch (InterruptedException | ExecutionException e) {
       throw new RuntimeException(e);
     } catch (TimeoutException e) {
-      System.out.println("Could not stop");
+      System.out.println("Could not stop, trying harder");
```
Comment (pierluca) [inline tests/karate/src/test/java/be/utils/Server.java:67]: 🤣

**Review by pierluca — APPROVED** [review]
LGTM 👍

**Pull request merged**